frontend/src/features/admin-form/create/builder-and-design/useFieldBuilderStore.tsx

```diff
@@ -11,64 +11,116 @@ export enum FieldBuilderState {
   Inactive,
 }
 
+type FieldBuilderCreateEditStateData =
+  | {
+      state: FieldBuilderState.CreatingField
+      field: FieldCreateDto
+      insertionIndex: number
+    }
+  | {
+      state: FieldBuilderState.EditingField
+      field: FormFieldDto
+    }
+  | { state: FieldBuilderState.EditingEndPage }
+
 export type FieldBuilderStore = {
-  updateCreateState: (field: FieldCreateDto, insertionIndex: number) => void
-  updateEditState: (field: FormFieldDto) => void
-  setEditEndPage: () => void
-  setToInactive: () => void
+  updateCreateState: (
+    field: FieldCreateDto,
+    insertionIndex: number,
+    holding?: boolean,
+  ) => void
+  updateEditState: (field: FormFieldDto, holding?: boolean) => void
+  setEditEndPage: (holding?: boolean) => void
+  setToInactive: (holding?: boolean) => void
+  isDirty: boolean
+  setIsDirty: (isDirty: boolean) => void
   stateData:
-    | {
-        state: FieldBuilderState.CreatingField
-        field: FieldCreateDto
-        insertionIndex: number
-      }
-    | {
-        state: FieldBuilderState.EditingField
-        field: FormFieldDto
-      }
-    | { state: FieldBuilderState.EditingEndPage }
+    | FieldBuilderCreateEditStateData
     | { state: FieldBuilderState.Inactive }
+  // Used when there is a dirty state and we want to hold the next state to be set.
+  // Will be used to set stateData if user confirms discarding changes.
+  holdingStateData:
+    | FieldBuilderCreateEditStateData
+    | { state: FieldBuilderState.Inactive }
```

Comment:
what's the reason for keeping the inactive state separate from `FieldBuilderCreateEditStateData`? didn't seem like there's anywhere that requires them to be separate

Reply:
reusing `FieldBuilderCreateEditStateData` type elsewhere where we want to ensure inactive state cannot be set.